---
title: "ComputingDivergnceWithRemovingN_06"
author: "Jesse Garcia"
date: "10/22/2018"
output: html_document
---


## not finished

```{r setup, include=FALSE, echo=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)

```



```{r cars, include=FALSE, eval=FALSE}


library(tidyverse)
library(data.table)
library(glue)
library(furrr)
library(lubridate)
library(GenomicRanges)
library(Biostrings)
vcfOfAlignments<-"../data/vcfOfAlignments_Oct19.csv"
vcfOfAlignments<-as_tibble(fread(vcfOfAlignments)) #%>% 
  #filter(chr=="chr5 Scq3eQI_2018;HRSCAF=3193")
```




## Newest estimate of absolute divergence is similar to old estimates  denominator and numerator

.02750342 (Removing whole Alignments  With N) vs .02756862 (No Ns at all )

```{r pressure, echo=FALSE, include=F}


filteredVCF<-vcfOfAlignments %>% 
  filter(qLobataSeq != "N", 
         qRoburSeq != "N", 
         qLobataSeq != "-", 
         qRoburSeq != "-")

matches<-sum(filteredVCF$qLobataSeq == filteredVCF$qRoburSeq)
mismatches<-sum(filteredVCF$qLobataSeq != filteredVCF$qRoburSeq)

mismatches/(matches+mismatches)
```


## Removing Repeats

```{r}
bedFileOfRepeatsAndN<-"../data/Qlobata3dot0-chromNamesJustFirstWord.runs-of-SJCdeNovoRptModelRptMask-and-Ns.UCSCbed4.txt"
bedFileOfRepeatsAndN<-fread(bedFileOfRepeatsAndN)
bedFileOfRepeatsAndN<-bedFileOfRepeatsAndN %>% dplyr::rename(IncompleteChrName=V1, Start=V2, End=V3, Annotation=V4)

#tables come from Sorel's email 
chrName<-read_table("chr1 Reversed: Scq3eQI_240;HRSCAF=820
chr2 Reversed: Scq3eQI_1869;HRSCAF=2693
chr3 Reversed: Scq3eQI_2027;HRSCAF=3423
chr4 Scq3eQI_2026;HRSCAF=3421
chr5 Scq3eQI_2018;HRSCAF=3193
chr6 Scq3eQI_2028;HRSCAF=3424
chr7 Reversed: Scq3eQI_27
chr8 Scq3eQI_103;HRSCAF=384
chr9 Reversed: Scq3eQI_316;HRSCAF=953
chr10 Scq3eQI_174;HRSCAF=633
chr11 Reversed: Scq3eQI_1982;HRSCAF=3004
chr12 Scq3eQI_304;HRSCAF=933", col_names=F)

chr<-read_table("chr1
chr2
chr3
chr4
chr5
chr6
chr7
chr8
chr9
chr10
chr11
chr12",col_names=F)

translationTable<-bind_cols(chrName, chr) %>% 
  dplyr::rename(QLobata2018Name=X1, IncompleteChrName=X11)


bedFileOfRepeatsAndN<-inner_join(  bedFileOfRepeatsAndN, translationTable, by=c("IncompleteChrName"="IncompleteChrName"))

##Preparing Bed File for Intersenction
repeatsAndNGRange<-GRanges(seqnames=bedFileOfRepeatsAndN$QLobata2018Name,IRanges(start=bedFileOfRepeatsAndN$Start, end=bedFileOfRepeatsAndN$End), BValue=bedFileOfRepeatsAndN$Annotation)


vcfOfAlignmentsGRange<-GRanges(seqnames=vcfOfAlignments$chr,IRanges(start=vcfOfAlignments$position, end=vcfOfAlignments$position))




hits<-findOverlaps(vcfOfAlignmentsGRange, repeatsAndNGRange)


indexOfRepeatsIntersections<-subjectHits(hits)
indexOfVCFIntersections<-queryHits(hits)

rm(hits)
rm(vcfOfAlignmentsGRange)
rm(repeatsAndNGRange)
intersectedVCF<-vcfOfAlignments[indexOfVCFIntersections,]

intersectedRepeats<-bedFileOfRepeatsAndN[indexOfRepeatsIntersections,]

rm(indexOfRepeatsIntersections)
rm(indexOfVCFIntersections)
rm(vcfOfAlignments)

intersectedVCF$Annotation<-intersectedRepeats$Annotation

fwrite(intersectedVCF, "intersectedVCF_Oct22.csv")



```



```{r}
intersectedVCF %>% filter(Annotation =="Nx1865")




intersectedVCF %>% filter(Annotation =="Nx1450")
intersectedVCF %>% filter(Annotation =="Nx446")
intersectedVCF %>% filter(Annotation =="Nx3044")

unique(intersectedVCF$Annotation)

#Doesnt work
intersectedVCF %>% filter(Annotation =="Nx388")






vcfOfAlignments %>% filter(position > 54634660 & position < 54634660 + 3)

bedFileOfRepeatsAndN %>% filter(Annotation == "Nx1753")


bedFileOfRepeatsAndN %>% filter(Annotation =="Nx388")

test=intersectedVCF %>% filter(Annotation != 	"RptMask(deNovo)")

mean(test$qLobataSeq == "N")
```



```{r}
qLobata2018Reference<-"../data/Qlobata.v3.0.mainChrs.fasta"


qLobata2018Reference<-readDNAStringSet(qLobata2018Reference)

dna<-subseq(qLobata2018Reference["chr1 Reversed: Scq3eQI_240;HRSCAF=820"], start=4361, end=4361+50)

as.character(dna)
```

