---
title: "Computing Oak Divergence Using MUMMER Alignments"
author: "Jesse Garcia"
date: "10/18/2018"
output: beamer_presentation
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(tidyverse)
library(data.table)
library(knitr)
library(kableExtra)
library(pander)
```

## Processing of Data

1. Used .coords file and Sorel's "chromosome name to scaffold and orientation table" to convert 2017 Q Lobata alignment coordinates to 2018 Q Lobata alignment coordinates
2. Removed all alignments that weren't on chromosomes
3. Used remaining alignment coordinates to extract nucleotide sequences from Q Robur and Q Lobata
4. Nucleotide sequences were differing lengths because of indels, so globally aligned sequence pairs using NW algorithm. Insertions were indicated with "_". Now Sequences are the same length and have 1-1 alignment.
5. Currently 128,513 Alignment pairs out of 129,162 finished computing (99.5%). 30 Pairs were too large for NW align. 



```{r cars, echo = FALSE, include=FALSE}
divergenceAlignments<-"../data/divergenceAlignments_Oct18.csv"
divergenceAlignments<-as_tibble(fread(divergenceAlignments))
divergenceAlignments<-divergenceAlignments %>% filter( !is.na(numberOfMatches)    | !is.na(numberOfMismatches)  )



divergenceAlignments[is.na(divergenceAlignments$numberOfMatches),]
```

## There's some variability of divergence across alignment coordinates

```{r warning=F,message=F}
ggplot(divergenceAlignments, aes(numberOfMismatches/totalLengthAligned)) + 
  geom_histogram() + 
  theme_bw() + 
  labs(y="Count", x="Species Divergence (Mismatches/Aligned)")
```



## I think runs of N's contribute to this
```{r warning=F, message=F}


divergenceAlignments<-divergenceAlignments %>% mutate(NPresent=str_detect(pattern, "N"))



divergenceAlignments %>% ggplot(aes(x=numberOfMismatches/totalLengthAligned, colour=NPresent)) +
  geom_density() + 
  theme_bw() + 
  labs(y="Count", x="Species Divergence (Mismatches/Aligned)", colour="N Present in Q Lobata") 
```


## Aligned N's Count as matches. Only last 6 bases are nucleotides. 
```{r}

divergenceAlignments$pattern[1]
divergenceAlignments$subject[1]
divergenceAlignments$numberOfMatches[1]

cat("NNNCAAATCT")
```




```{r include=F}
totalLengthAligned<-sum(divergenceAlignments$numberOfMatches + divergenceAlignments$numberOfMismatches, na.rm = T)
totalNumberOfMismatches<-sum(divergenceAlignments$numberOfMismatches, na.rm = T)
divergenceForAll<-totalNumberOfMismatches/totalLengthAligned

```


## Divergence Estimates by N

Total Divergence for everything: `r divergenceForAll`

```{r}





table<-divergenceAlignments %>% 
  dplyr::group_by(NPresent) %>% 
  summarise(Divergence=sum(numberOfMismatches, na.rm=T) /sum(numberOfMatches + numberOfMismatches, na.rm=T), 
            Alignments=n(),
            TotalLength=sum(totalLengthAligned, na.rm=T),
            NumberOfMatches=sum(numberOfMatches, na.rm=T),
            NumberOfMismatches=sum(numberOfMismatches, na.rm=T))

pander(table)
```




## Next Steps
1. Remove Repeats, SNPs near Runs of Ns, and CDS.
2. Sorel also has GATK called variants where Q Robur and Q Lobata were aligned to a reference genome. I think Computing divergence for this should be easier because scripts exist. 


```{r}
divergenceAlignments %>% 
  dplyr::group_by(NPresent) %>% 
  summarise(Divergence=sum(numberOfMismatches, na.rm=T) /sum(numberOfMatches + numberOfMismatches, na.rm=T), 
            Alignments=n(),
            TotalLength=sum(totalLengthAligned, na.rm=T),
            NumberOfMatches=sum(numberOfMatches, na.rm=T),
            NumberOfMismatches=sum(numberOfMismatches, na.rm=T))

```




```{r include=FALSE             }
divergenceAlignments %>% ggplot(aes(x=totalLengthAligned, y=(numberOfMismatches/totalLengthAligned), colour=NPresent)) + geom_point() + theme_bw()


divergenceAlignments %>% ggplot(aes(x=totalLengthAligned)) + geom_histogram()


summary(divergenceAlignments$totalLengthAligned/abs(divergenceAlignments$QLobata2018End - divergenceAlignments$QLobata2018Start))

summary(divergenceAlignments$totalLengthAligned)
```

