---
title: "R Notebook"
output: html_notebook
---

 


```{r}
library(tidyverse)
library(data.table)
library(GenomicRanges)
library(Biostrings)
```




#Reading in data


```{r}
snps<-"../data/QrobToQlobNucmer/oak_Qr.snps"
snps<-as_tibble(fread(snps))


# Under show-snps at this site http://mummer.sourceforge.net/manual/
snps<-snps %>% dplyr::rename(QLobata2017Position=V1,
                QLobataAllele=V2,
                QRoburAllele=V3,
                QRoburPosition=V4,
                DistanceToNearestMismatch=V5,
                DistanceToNearestSequenceEnd=V6,
                LengthOfQLobataSequence=V7,
                LengthOfQRoburSequence=V8,
                QLobataSequenceDirection=V9,
                QRoburSequenceDirection=V10,
                QLobata2017Names=V11,
                QRoburNames=V12)



repeatsAndN<-"../data/Qlobata3dot0-chromNamesJustFirstWord.runs-of-SJCdeNovoRptModelRptMask-and-Ns.UCSCbed4.txt"
repeatsAndN<-as_tibble(fread(repeatsAndN))


repeatsAndN<-repeatsAndN %>% 
  dplyr::rename(IncompleteChromosomeNames=V1, QLobata2018Start=V2, 
         QLobata2018End=V3, Annotation=V4)




wholeCoordinatesAndAlignments<-"../data/CoordinatesAndAlignmentsForQRoburAndQLobata_Oct172018.csv"


#CSV has "" which represent NA. Change all "" to NA
wholeCoordinatesAndAlignments<-as_tibble(fread(wholeCoordinatesAndAlignments)) %>%
  mutate_all(funs(na_if(., "")))



coordinatesAndAlignments<-wholeCoordinatesAndAlignments %>% 
  select(QLobata2018Start,QLobata2018End, QLobata2018Chromosome, QRoburStart, QRoburEnd, QLobata2018Names)


coordinatesAndAlignments<-coordinatesAndAlignments %>%
  mutate(IncompleteChromosomeNames=str_extract(QLobata2018Chromosome, "chr[0-9]*"))

translation<-coordinatesAndAlignments %>% 
  select(IncompleteChromosomeNames,QLobata2018Names) %>% 
  unique()



repeatsAndN<-left_join(repeatsAndN,translation)





```




# Checking to see whether all Alignments come from chromosomes, and whether all Chromosomes have alignments
```{r}
wholeCoordinatesAndAlignments  %>% filter(!is.na(QLobata2018SeqAlignments)) %>%
  pull(QLobata2018Names) %>%
  table()


coordinatesAndAlignments %>% filter(!is.na(QLobata2018Chromosome)) %>%
  pull(QLobata2018Chromosome) %>%
  table()


```



# Getting only the SNPs in Chromosome alignments. 

First creating  new variables QLobata2018Names and QLobata2018Chromosome that contains the translation from the Q Lobata 2017 Genome to the Q Lobata 2018 Genome
```{r}
translation <- wholeCoordinatesAndAlignments %>% 
  select(QLobata2017Names, QLobata2018Names, QLobata2018Chromosome,Length,Reversed) %>% 
  unique() %>% 
  filter(!is.na(QLobata2018Chromosome)) 

# Chromosome 7 needs to be replaced with "Scq3eQI_27;HRSCAF=100"
translation$QLobata2017Names<-case_when(
translation$QLobata2017Names== "Scq3eQI_27" ~  "Scq3eQI_27;HRSCAF=100" ,
TRUE ~ translation$QLobata2017Names
)


translation<-translation %>%
  mutate(Reversed=str_detect(QLobata2018Chromosome,"Reversed"))


#Only SNPs and Indels on the Chromosomes are kepts. 
snps=inner_join(snps, translation,  by="QLobata2017Names")

#rm(wholeCoordinatesAndAlignments)

```




# Testing new thing (THIS IS A MESS)

Aligning sequences together globaly using the Needleman-Wunsch global alignment. 


```{r}

library(tidyverse)
library(data.table)
library(GenomicRanges)
library(Biostrings)
library(furrr)


plan(multiprocess)

wholeCoordinatesAndAlignments<-"../data/CoordinatesAndAlignmentsForQRoburAndQLobata_Oct172018.csv"


#CSV has "" which represent NA. Change all "" to NA
wholeCoordinatesAndAlignments<-as_tibble(fread(wholeCoordinatesAndAlignments)) %>%
  mutate_all(funs(na_if(., ""))) %>% 
  filter(!is.na(QLobata2018SeqAlignments))




#sampledCoordinatesAndAlignments<-wholeCoordinatesAndAlignments %>% sample_n(100)


sampledCoordinatesAndAlignments<-wholeCoordinatesAndAlignments %>% sample_n(100)


getValuesNeededForDivergence<-function(qLobataSequence, qRoburSequence, qLobata2018Start, qLobata2018Name, qLobata2018End){
pairAlign=pairwiseAlignment(pattern=qLobataSequence,subject=qRoburSequence, type="global")

numberOfMatches<-nmatch(pairAlign)
numberOfMismatches<-nmismatch(pairAlign)
#mismatch<-as_tibble(mismatchTable(pairAlign))
valuesNeededForDivergence<-tibble(
  numberOfMatches=numberOfMatches,
  numberOfMismatches=numberOfMismatches, 
  pattern=as.character(pairAlign@pattern),
  subject=as.character(pairAlign@subject),
  QLobata2018Start=qLobata2018Start,
  QLobata2018Name=qLobata2018Name,
  QLobata2018End=qLobata2018End,
  totalLengthAligned=(numberOfMatches+numberOfMismatches)
)
return(valuesNeededForDivergence)
}


possibly_getValuesNeededForDivergence<-possibly(getValuesNeededForDivergence, tibble(numberOfMatches=NA,numberOfMismatches=NA, pattern=NA, subject=NA, QLobata2018Start=NA, QLobata2018Name=NA, QLobata2018End=NA ) )



l <- list(qLobataSequence = sampledCoordinatesAndAlignments$QLobata2018SeqAlignments,
          qRoburSequence = sampledCoordinatesAndAlignments$QRoburSeqAlignments, 
          qLobata2018Start = sampledCoordinatesAndAlignments$QLobata2018Start, 
          qLobata2018Name=sampledCoordinatesAndAlignments$QLobata2018Names, 
          qLobata2018End=sampledCoordinatesAndAlignments$QLobata2018End)

valuesNeededForDivergence=pmap(l,possibly_getValuesNeededForDivergence)



#valuesNeededForDivergence=future_map2(sampledCoordinatesAndAlignments$QLobata2018SeqAlignments,sampledCoordinatesAndAlignments$QRoburSeqAlignments, ~ possibly_getValuesNeededForDivergence(.x,.y), .progress=T)

valuesNeededForDivergence<-valuesNeededForDivergence %>% bind_rows()

saveRDS(valuesNeededForDivergence, "../data/valuesNeededForDivergence_Oct172018Mac.rds")

```





```{r}
pairAlign=pairwiseAlignment(pattern=qLobataSequence,subject=qRoburSequence, type="global")




```






# Converting SNPs 2017 Position to 2018 Position

Using the same alignment rules in OakTreeAlignments_01.Rmd we will create a new variable QLobata2018Position.

```{r}





#If Reversed: Length - Position + 1
snpsReversed<-snps %>% 
  filter(Reversed == TRUE) %>%
  mutate(QLobata2018Position=Length-QLobata2017Position+1) 

#If Not reversed :just use the original Position
snpsNotReversed<-snps %>% 
  filter(Reversed == FALSE) %>%
  mutate(QLobata2018Position=QLobata2017Position) 


snps<-bind_rows(snpsReversed,snpsNotReversed)



```


#Testing SNP Calling

```{r}
qLobata2018Reference<-"../data/Qlobata.v3.0.mainChrs.fasta"
qRoburReference<-"../data/Qrob_PM1N.fa"
qLobata2018Reference<-readDNAStringSet(qLobata2018Reference)
qRoburReference<-readDNAStringSet(qRoburReference)



getSequenceFromReference<-function(referenceGenome, scaffoldName, start, end){
  #If start is bigger than end, I think aligns to other strand. Change start and end and 
  # Reverse complement
  if (start > end){
    temp<-start
    start<-end
    end<-temp
    dna<-subseq(referenceGenome[scaffoldName], start=start, end=end)
    reversedDNA<-reverseComplement(dna)
    return(reversedDNA)
  } else {
    dna<-subseq(referenceGenome[scaffoldName], start=start, end=end)
    return(dna)
  }
  
  
}
possibly_getSequenceFromReference<-possibly(getSequenceFromReference, otherwise=NA )



sampleOfSnps<-snps %>% dplyr::slice(30:41)



QRoburSeqAlignments<-sampleOfSnps %>% rowwise() %>% do(
  as_tibble(as.character(getSequenceFromReference(referenceGenome = qRoburReference, 
                           scaffoldName = .$QRoburNames,
                           start=.$QRoburPosition,
                           end=.$QRoburPosition+10)))
) %>% dplyr::rename(QRoburSeqAlignments=value)


QLobata2018SeqAlignments<-sampleOfSnps %>% rowwise() %>% do(
  as_tibble(as.character(possibly_getSequenceFromReference(referenceGenome = qLobata2018Reference,                   scaffoldName = .$QLobata2018Names,
                           start=.$QLobata2018Position,
                           end=.$QLobata2018Position+10)))
)  %>% dplyr::rename(QLobata2018SeqAlignments=value)

test=bind_cols(QRoburSeqAlignments,QLobata2018SeqAlignments)


subseq(qLobata2018Reference[snps$QLobata2018Names[1]], 
       start=snps$QLobata2018Position[1], 
       end=snps$QLobata2018Position[1]+3)




subseq(qRoburReference[snps$QRoburNames[1]], 
       start=snps$QRoburPosition[1], 
       end=snps$QRoburPosition[1]+3)

snps$QRoburAllele[1]

snps$QLobataAllele[1]
```




#Making sure all chromsomes have SNPs and that SNPs are only on chromosomes. 
```{r}
table(snps$QLobata2018Chromosome)



```

# Indels 
```{r}
#Number of Total Indels
snps %>% filter(QLobataAllele == "." |  QRoburAllele == ".") %>% nrow()

#Number Of times Q Lobata has "." While Q Robur has a Nucleotide. Q Robur has an insertion. 
snps %>% filter(QLobataAllele == ".") %>% nrow()


#Number Of times Q Robur has "." While Q Lobata has a Nucleotide. Q Lobata has an insertion. 
snps %>% filter(QRoburAllele == ".") %>% nrow()
```


# Filtering out SNPs in repeats. 
```{r}

repeatsAndN<-repeatsAndN %>% filter(!is.na(QLobata2018Names)) 


repeatsRanges = IRanges(start=repeatsAndN$QLobata2018Start, end = repeatsAndN$QLobata2018End)
repeatsAndNRange<-GRanges(seqnames = repeatsAndN$QLobata2018Names, repeatsRanges)


snpsRanges<-IRanges(start=)



```





