---
title: "R Notebook"
output: html_notebook
---


Trying to recreate yings filters....
```{r}
knitr::opts_chunk$set(echo = FALSE)

knitr::opts_chunk$set(dev = 'pdf')

library(data.table)
library(tidyverse)
library(glue)
library(bedr)


```



## Step 1. Remove indels, multiallelics, remove repeats. Run on Hoffman

vcftools --gzvcf data_from_sorel/24genomes/2018wgs2.ef.vcf.gz --remove-indels --recode --recode-INFO-all --chr chr${SGE_TASK_ID} --exclude-bed ~/klohmueldata/ying_data/oak_genome/data_from_sorel/Qlobata.v3.0.repeats.bed --min-alleles 1 --max-alleles 2 --out 2018wgs2.ef.rmIndelRepeats.chr${SGE_TASK_ID}



original Resequenced "/u/flashscratch/z/zhen/forJesse/resequenced/"
module load R/3.5.0
module load zlib/1.2.8
module load vcftools

```{r}
library(tidyverse)
library(glue)

vcftools<-"vcftools"
resequencedVCF<-"/u/flashscratch/j/jessegar/forJesse/data_from_sorel/2018wgs2.ef.vcf.gz"
qLobataRepeatsBed<-"/u/flashscratch/j/jessegar/forJesse/data_from_sorel/Qlobata.v3.0.repeats.bed"
outputDir<-"/u/flashscratch/j/jessegar/TryingToRecreateYingsFilters"

filters<-tibble(
  chromosome=1:12,
  vcftools=vcftools,
  resequencedVCF=resequencedVCF,
  qLobataRepeatsBed=qLobataRepeatsBed
)


filters<-filters %>% 
  mutate(output=glue("{outputDir}/2018wgs2.ef.rmIndelRepeats.chr{chromosome}")) %>%
  mutate(command=glue("{vcftools} --gzvcf {resequencedVCF} --remove-indels --recode --recode-INFO-all --chr chr{chromosome} --exclude-bed {qLobataRepeatsBed} --min-alleles 1 --max-alleles 2 --out {output}"))







system(filters$command[12])







```


## Removing upstream Indels
vcftools --vcf 2018wgs2.ef.rmIndelRepeatsStar.chr$i.vcf --indv ${sample_array[$SGE_TASK_ID-1]} --min-meanDP 12 --recode --recode-INFO-all --out ${sample_array[$SGE_TASK_ID-1]}.2018wgs2.ef.rmIndelRepeatsStar.chr$i.minDP12



```{r}

filters<-filters %>% 
  mutate(commandToRemoveUpstreamIndels=glue("grep -vw '\\*' {output}.recode.vcf > /u/flashscratch/j/jessegar/TryingToRecreateYingsFilters/2018wgs2.ef.rmIndelRepeatsStar.chr{chromosome}.vcf" )   ) 
  




```




## Extracting individual
vcftools --vcf 2018wgs2.ef.rmIndelRepeatsStar.chr$i.vcf --indv ${sample_array[$SGE_TASK_ID-1]} --min-meanDP 12 --recode --recode-INFO-all --out ${sample_array[$SGE_TASK_ID-1]}.2018wgs2.ef.rmIndelRepeatsStar.chr$i.minDP12

extract single indiv and filter by min depth 12x





```{r}
vcftools<-"vcftools"
resequencedVCF<-"/u/flashscratch/j/jessegar/forJesse/data_from_sorel/2018wgs2.ef.vcf.gz"
qLobataRepeatsBed<-"/u/flashscratch/j/jessegar/forJesse/data_from_sorel/Qlobata.v3.0.repeats.bed"
outputDir<-"/u/flashscratch/j/jessegar/TryingToRecreateYingsFilters"

filters<-tibble(
  chromosome=1:12,
  vcftools=vcftools,
  individual="QL.CR12.6.00F"
)

filters<-filters %>%
  mutate(input=glue("{outputDir}/2018wgs2.ef.rmIndelRepeatsStar.chr{chromosome}.vcf")) %>%
  mutate(output=glue("{outputDir}/{individual}.2018wgs2.ef.rmIndelRepeatsStar.chr{chromosome}.minDP12"  ))



filters<-filters %>% 
  mutate(extractingIndividual=glue("{vcftools} --vcf {input} --indv {individual} --min-meanDP 12 --recode --recode-INFO-all --out {output}" )   )  
  

system(filters$extractingIndividual[12])



```



## Convert this vcf to bed file for the masks

module load bedtools/2.26.0

vcf_coordinates | awk 'BEGIN{OFS="\t"} {print $1, $2-1, $2}' | sort -k1,1 -k2,2n | bedtools merge -i stdin > out_bedFile


awk '{{OFS="\\t"; if (!/^#/){print $1,$2-1,$2,$4"/"$5,"+"}}}'

QL.CR17.1.00F.2018wgs2.ef.rmIndelRepeatsStar.chr12.minDP12.callable.bed 
```{r}

vcftools<-"vcftools"
bedtools<-"bedtools"
qLobataRepeatsBed<-"/u/flashscratch/j/jessegar/forJesse/data_from_sorel/Qlobata.v3.0.repeats.bed"
outputDir<-"/u/flashscratch/j/jessegar/TryingToRecreateYingsFilters"

filters<-tibble(
  chromosome=1:12,
  vcftools=vcftools,
  individual="QL.CR12.6.00F"
)


x=read.vcf("/u/flashscratch/j/jessegar/TryingToRecreateYingsFilters/QL.CR12.6.00F.2018wgs2.ef.rmIndelRepeatsStar.chr12.minDP12.recode.vcf")
vcf2bed(x, filename = "/u/flashscratch/j/jessegar/TryingToRecreateYingsFilters/QL.CR12.6.00F.2018wgs2.ef.rmIndelRepeatsStar.chr12.minDP12.callable.bed")
bed<-as_tibble(fread("/u/flashscratch/j/jessegar/TryingToRecreateYingsFilters/QL.CR12.6.00F.2018wgs2.ef.rmIndelRepeatsStar.chr12.minDP12.callable.bed"))
bed$V3 <- bed$V2 + 1
bedMerged=as_tibble(bedr.merge.region(bed, check.chr = FALSE,check.valid = FALSE,check.sort = FALSE, check.zero.based = FALSE, list.names = FALSE))

write_tsv(bedMerged, path="/u/flashscratch/j/jessegar/TryingToRecreateYingsFilters/QL.CR12.6.00F.2018wgs2.ef.rmIndelRepeatsStar.chr12.minDP12.callable.bed",col_names=FALSE)

```


## 
